src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -1931,4 +1940,215 @@ public void Dispose()
+
+        private Dictionary<object, DocumentInfo> _onBeforeStoreDocumentsByEntity;
+
+        private bool _prepareEntitiesPuts;
+
+        public int Count => _documentsByEntity.Count + _onBeforeStoreDocumentsByEntity?.Count ?? 0;
+
+        public void Remove(object entity)
+        {
+            _documentsByEntity.Remove(entity);
+            _onBeforeStoreDocumentsByEntity?.Remove(entity);
+        }
+
+        public void Evict(object entity)
+        {
+            if (_prepareEntitiesPuts)
+                throw new InvalidOperationException("Cannot Evict entity during OnBeforeStore");
+
+            _documentsByEntity.Remove(entity);
+        }
+
+        public void Add(object entity, DocumentInfo documentInfo)
+        {
+            if (_prepareEntitiesPuts == false)
+            {
+                _documentsByEntity.Add(entity, documentInfo);
+                return;
+            }
+
+            CreateOnBeforeStoreDocumentsByEntityIfNeeded();
+            _onBeforeStoreDocumentsByEntity.Add(entity, documentInfo);
+        }
+
+        public DocumentInfo this[object obj]
+        {
+            set
+            {
+                if (_prepareEntitiesPuts == false)
+                {
+                    _documentsByEntity[obj] = value;
+                }
+                else
+                {
+                    CreateOnBeforeStoreDocumentsByEntityIfNeeded();
+                    _onBeforeStoreDocumentsByEntity[obj] = value;
+                }
+            }
+        }
+
+        private void CreateOnBeforeStoreDocumentsByEntityIfNeeded()
+        {
+            if (_onBeforeStoreDocumentsByEntity != null)
+                return;
+
+            _onBeforeStoreDocumentsByEntity = new Dictionary<object, DocumentInfo>(ObjectReferenceEqualityComparer<object>.Default);
+        }
+
+        public void Clear()
+        {
+            _documentsByEntity.Clear();
+            _onBeforeStoreDocumentsByEntity?.Clear();
+        }
+
+        public bool TryGetValue(object entity, out DocumentInfo documentInfo)
+        {
+            if (_documentsByEntity.TryGetValue(entity, out documentInfo))
+                return true;
+
+            return _onBeforeStoreDocumentsByEntity != null && _onBeforeStoreDocumentsByEntity.TryGetValue(entity, out documentInfo);
+        }
+
+        public IEnumerator<DocumentsByEntityEnumeratorResult> GetEnumerator()
+        {
+            foreach (var doc in _documentsByEntity)
+            {
+                yield return new DocumentsByEntityEnumeratorResult
+                {
+                    Key = doc.Key,
+                    Value = doc.Value,
+                    ExecuteOnBeforeStore = true
+                };
+            }
+
+            if (_onBeforeStoreDocumentsByEntity != null)
+            {
+                foreach (var doc in _onBeforeStoreDocumentsByEntity)
```

Comment:
Create a new instance, and null the old one.
That way, we can get call on our own changes and safely handle them

Reply:
We need the old one since we track those entities